rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala

```diff
@@ -21,66 +22,74 @@ object Dispatch {
 }
 
-class RholangOnlyDispatcher private (_reducer: => Reduce[Task])
-    extends Dispatch[Task, Seq[Channel], TaggedContinuation] {
+class RholangOnlyDispatcher[M[_]] private (_reducer: => Reduce[M])(implicit captureM: Capture[M])
+    extends Dispatch[M, Seq[Channel], TaggedContinuation] {
 
-  val reducer: Reduce[Task] = _reducer
+  val reducer: Reduce[M] = _reducer
 
-  def dispatch(continuation: TaggedContinuation, dataList: Seq[Seq[Channel]]): Task[Unit] =
+  def dispatch(continuation: TaggedContinuation, dataList: Seq[Seq[Channel]]): M[Unit] =
     continuation.taggedCont match {
       case ParBody(par) =>
         val env = Dispatch.buildEnv(dataList)
         reducer.eval(par)(env)
       case ScalaBodyRef(_) =>
-        Task.unit
+        captureM.apply(())
       case Empty =>
-        Task.unit
+        captureM.apply(())
     }
 }
 
 object RholangOnlyDispatcher {
 
-  def create(tuplespace: IStore[Channel, BindPattern, Seq[Channel], TaggedContinuation])
-    : Dispatch[Task, Seq[Channel], TaggedContinuation] = {
-    lazy val dispatcher: Dispatch[Task, Seq[Channel], TaggedContinuation] =
+  def create[M[_], F[_]](
+      tuplespace: IStore[Channel, BindPattern, Seq[Channel], TaggedContinuation])(
+      implicit
+      intepreterErrorsM: InterpreterErrorsM[M],
+      captureM: Capture[M],
+      parallel: Parallel[M, F]): Dispatch[M, Seq[Channel], TaggedContinuation] = {
+    lazy val dispatcher: Dispatch[M, Seq[Channel], TaggedContinuation] =
       new RholangOnlyDispatcher(reducer)
-    lazy val reducer: Reduce[Task] =
-      new Reduce.DebruijnInterpreter[Task, Task.Par](tuplespace, dispatcher)
+    lazy val reducer: Reduce[M] =
+      new Reduce.DebruijnInterpreter[M, F](tuplespace, dispatcher)
     dispatcher
   }
 }
 
-class RholangAndScalaDispatcher private (
-    _reducer: => Reduce[Task],
-    _dispatchTable: => Map[Long, Function1[Seq[Seq[Channel]], Task[Unit]]])
-    extends Dispatch[Task, Seq[Channel], TaggedContinuation] {
+class RholangAndScalaDispatcher[M[_]] private (
+    _reducer: => Reduce[M],
+    _dispatchTable: => Map[Long, Function1[Seq[Seq[Channel]], M[Unit]]])(
+    implicit captureM: Capture[M])
+    extends Dispatch[M, Seq[Channel], TaggedContinuation] {
 
-  val reducer: Reduce[Task] = _reducer
+  val reducer: Reduce[M] = _reducer
 
-  def dispatch(continuation: TaggedContinuation, dataList: Seq[Seq[Channel]]): Task[Unit] =
+  def dispatch(continuation: TaggedContinuation, dataList: Seq[Seq[Channel]]): M[Unit] =
     continuation.taggedCont match {
       case ParBody(par) =>
         val env = Dispatch.buildEnv(dataList)
         reducer.eval(par)(env)
       case ScalaBodyRef(ref) =>
         _dispatchTable.get(ref) match {
           case Some(f) => f(dataList)
-          case None    => Task.raiseError(new Exception(s"dispatch: no function for $ref"))
+          case None    => captureM.fail(new Exception(s"dispatch: no function for $ref"))
```

Comment:
Is is equivalent to `MonadError[M].raiseError` ?

Reply:
Again, it *might be* (though I have to wonder about `MonadError` instances for `Task` given the fact that `Task` isn't parametric in its error type).

Again, in my mind, this a matter of a more faithful translation from the original implementation with a given concrete monad (`Task`) to the equivalent type class (`Capture`).

Reply:
Yeah.  Seems like there's going to need to be some attention paid to `Sync` vs `Async` stuff WRT to our type classes.  I'm going with `Capture` because it seems like the right incremental move right now.